core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -753,10 +753,17 @@ public static Jenkins getInstanceOrNull() {
      * @throws IllegalStateException {@link Jenkins} has not been started, or was already shut down
      */
     @CLIResolver
-    @Nullable // TODO replace with non-null once Jenkins 2.0+
+    @Nonnull
     public static Jenkins getInstance() {
-        // TODO throw an IllegalStateException in Jenkins 2.0+
-        return HOLDER.getInstance();
+        Jenkins instance = HOLDER.getInstance();
+        if (instance == null) {
+            if(!Boolean.getBoolean(Jenkins.class.getName()+".disableExceptionOnNullInstance")) {
+                // remove that second block around 2.20 (that is: ~20 versions to battle test it)
+                // See https://github.com/jenkinsci/jenkins/pull/2297#issuecomment-216710150
+                throw new IllegalStateException();
```

Comment:
Can you elaborate @oleg-nenashev ? Not sure which original text you're talking about. 

`// TODO throw an IllegalStateException in Jenkins 2.0+` ?